### PR TITLE
[FINERACT-1760] Client transaction by transaction external ID

### PR DESCRIPTION
**Pull request by taskain7**
## Description

 - Pay charge with transaction external ID
 - retrieve client transaction by transaction external ID
 - undo client transaction by transaction external ID

### CODE REVIEW

**Review by adamsaghy — CHANGES_REQUESTED** [review]
@taskain7 Please rebase this PR, there is a merge commit!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/ClientTransaction.java [hunk 1]
```diff
@@ -98,24 +99,23 @@ public class ClientTransaction extends AbstractAuditableWithUTCDateTimeCustom {
...
-        final String externalId = null;
         return new ClientTransaction(client, office, paymentDetail, ClientTransactionType.PAY_CHARGE.getValue(), transactionDate, amount,
                 isReversed, externalId, currencyCode);
     }
 
     public static ClientTransaction waiver(final Client client, final Office office, final LocalDate transactionDate, final Money amount,
             final String currencyCode) {
         final boolean isReversed = false;
-        final String externalId = null;
+        final ExternalId externalId = null;
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/ClientTransaction.java]: ExternalId by design should never be null. Please use ExternalId.empty() in this case.

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see my comment!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/ClientTransaction.java (hunk 1 above)
Comment (taskain7) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/ClientTransaction.java]: fixed

**Review by adamsaghy — APPROVED** [review]
LGTM

**Pull request merged**